src/components/views/rooms/BasicMessageComposer.js

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2019 New Vector Ltd
+Copyright 2019 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import React from 'react';
+import {_t} from '../../../languageHandler';
+import PropTypes from 'prop-types';
+import dis from '../../../dispatcher';
+import EditorModel from '../../../editor/model';
+import HistoryManager from '../../../editor/history';
+import {setCaretPosition} from '../../../editor/caret';
+import {getCaretOffsetAndText} from '../../../editor/dom';
+import Autocomplete from '../rooms/Autocomplete';
+import {autoCompleteCreator} from '../../../editor/parts';
+import {renderModel} from '../../../editor/render';
+import {Room} from 'matrix-js-sdk';
+
+const IS_MAC = navigator.platform.indexOf("Mac") !== -1;
+
+export default class BasicMessageEditor extends React.Component {
+    static propTypes = {
+        model: PropTypes.instanceOf(EditorModel).isRequired,
+        room: PropTypes.instanceOf(Room).isRequired,
+    };
+
+    constructor(props, context) {
+        super(props, context);
+        this.state = {
+            autoComplete: null,
+        };
+        this._editorRef = null;
+        this._autocompleteRef = null;
+        this._modifiedFlag = false;
+    }
+
+    _updateEditorState = (caret, inputType, diff) => {
+        renderModel(this._editorRef, this.props.model);
+        if (caret) {
+            try {
+                setCaretPosition(this._editorRef, this.props.model, caret);
+            } catch (err) {
+                console.error(err);
+            }
+        }
+        this.setState({autoComplete: this.props.model.autoComplete});
+        this.historyManager.tryPush(this.props.model, caret, inputType, diff);
```

Comment:
Probably yes... it's still good to store the undo text, even if the caret was invalid. When undoing, setting the caret will indeed probably fail again, but at least you'll have the correct text.

The try/catch is to catch any bugs in the caret setting DOM code, which I haven't seen happening recently. Keeping it there as continuing is still the best thing to do if it would happen.